src/main/scala/mesosphere/marathon/raml/AppConversion.scala

```diff
@@ -323,7 +323,7 @@ trait AppConversion extends DefaultConversions with ConstraintConversion with En
       args = if (service.hasCmd && service.getCmd.getArgumentsCount > 0) service.getCmd.getArgumentsList.to[Seq] else App.DefaultArgs,
       backoffFactor = service.whenOrElse(_.hasBackoffFactor, _.getBackoffFactor, App.DefaultBackoffFactor),
       backoffSeconds = service.whenOrElse(_.hasBackoff, b => (b.getBackoff / 1000L).toInt, App.DefaultBackoffSeconds),
-      cmd = if (service.hasCmd && service.getCmd.hasValue) Option(service.getCmd.getValue) else App.DefaultCmd,
+      cmd = if (service.hasCmd && service.getCmd.getArgumentsCount == 0 && service.getCmd.hasValue) Option(service.getCmd.getValue) else App.DefaultCmd,
```

Comment:
I'm not sure if this is the correct check. The conversion should be "dumb" and just set what's in the Protobuf. E.g. if the protobuf defined arguments and a command then the command would be ignored during the conversion. I don't think we want that.

Reply:
The conversion is not so dumb. [Marathon protobuf definition of Service](https://github.com/mesosphere/marathon/blob/54b1f2611c90eafea967133dcd4bcb4eb75b583a/src/main/proto/marathon.proto#L109-L111) include Mesos's CommandInfo proto, [which semantic](https://github.com/mesosphere/marathon/blob/54b1f2611c90eafea967133dcd4bcb4eb75b583a/src/main/proto/mesos/mesos.proto#L682-L697) is quite different from Marathon's cmd and args.

What I mean is that it is perfectly fine for Marathon to enforce an excluding validation rule between cmd and args, but the storage format is too different so there is a need for a little more logic than dumb to implement conversion here.

As I said in my commit message, this is a tricky thing that has already been implemented in Marathon, [where the protobuf data are read from the storage and converted into an AppDefinition object](https://github.com/mesosphere/marathon/blob/54b1f2611c90eafea967133dcd4bcb4eb75b583a/src/main/scala/mesosphere/marathon/state/AppDefinition.scala#L222-L228). The issue  I propose to fix here is, to my knowledge, restricted to the 1.5.0 storage migration code, which [reimplemented the aforementioned logic](https://github.com/mesosphere/marathon/blob/54b1f2611c90eafea967133dcd4bcb4eb75b583a/src/main/scala/mesosphere/marathon/raml/AppConversion.scala#L323-L326) with a regression when it comes to args-only application.

Reply:
You are right. `cmd` and `args` are set in what we save, see [TaskBuilder.commandInfo](https://github.com/mesosphere/marathon/blob/releases/1.4/src/main/scala/mesosphere/mesos/TaskBuilder.scala#L241). @jdef verified that it's correct.

AFAICT, you are also right, that this conversion has to have some logic.

Now, the question is whether your line is correct. I'd say we compare `args[0]`  and `cmd`. If they are equal we drop `cmd`. What do you think?

Reply:
Alright, I've wrote a small test that verifies your fix. It fails on master. Could you merge [karsten/migration-error](https://github.com/mesosphere/marathon/tree/karsten/migration-error) into your branch?

So now we have to decide on what condition we set `cmd` or drop it. @jdef, what do you think?

Reply:
@jeschkies sorry for the Migration15 test you asked for, I tried an implementation yesterday but it was always passing on my side (even without my patch appied). I'll try with your test and let you know.

Reply:
@jeschkies I confirm your test is working: it gets green with my patch, a it fails when my patch is not applied:
```MigrationTo1_5 when migrating a single app should
- without cmd and container but with args defined *** FAILED *** (48 milliseconds)
  The future returned an exception of type: mesosphere.marathon.ValidationFailedException, with message: Validation failed: Failure(Set(value with value "App(/foo,None,Vector(sleep, 42),1.15,1,Some(sleep),Set(),None,0.1,Set(),10.0,Map(),//cmd,List(),Set()
,1,Map(),3600,64.0,0,None,Vector(Network(None,host,Map())),None,Some(Vector(PortDefinition(80,Map(),None,tcp))),List(),None,false,Map(),None,Some(UpgradeStrategy(1.0,1.0)),None,None,Some(1970-01-01T00:00:01Z),None,YOUNGEST_FIRST,Some(UnreachableEnabled(0,
0)),None)" AppDefinition must either contain one of 'cmd' or 'args', and/or a 'container'.)). (MigrationTo15Test.scala:29)
...```

I'll merge your test in my branch, but I have to exclude some unrelated commits.

Reply:
this looks correct to me and seems to mirror the `mergeFromProto` logic: only set the cmd if there are no args (because if there are args then cmd is argv[0] and should be ignored when converting from proto to RAML). maybe a comment here would be nice:
```scala
// Special consideration for cmd in the presence of args because we're converting from a Mesos-normalized 
// CommandInfo which has different semantics for cmd and args than App RAML. See AppDefinition.mergeFromProto.
```

Reply:
Awesome. I triggered the build. Now it's up to James.